### PR TITLE
Improve performance of information schema query.

### PR DESCRIPTION
**Pull request by arthurschreiber**
## Description

Querying the `information_schema` tables can be quite expensive in MySQL 5.7.

On the test database I'm using (169 tables), each execution of the `TablesWithSize57` query takes multiple seconds, making this query unbearably slow.

But the runtime of this query can be improved with a few tricks.

For reference, here's the original query:

```sql
	SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, sum(i.file_size), sum(i.allocated_size) 
	FROM information_schema.tables t, information_schema.innodb_sys_tablespaces i 
	WHERE t.table_schema = database() and 
	(i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')) 
	group by t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size
UNION ALL
	SELECT table_name, table_type, unix_timestamp(create_time), table_comment, SUM( data_length + index_length), SUM( data_length + index_length)
	FROM information_schema.tables t
	WHERE table_schema = database() AND 
	NOT EXISTS(SELECT * FROM information_schema.innodb_sys_tablespaces i WHERE i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')) 
	group by table_name, table_type, unix_timestamp(create_time), table_comment
```

```
169 rows in set (5.36 sec)
```

The first improvement is to reduce the number of accesses to the `information_schema.innodb_sys_tablespaces` table by replacing the `NOT EXISTS` part of the query with a `LEFT OUTER JOIN`:

```sql
	SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, sum(i.file_size), sum(i.allocated_size) 
	FROM information_schema.tables t, information_schema.innodb_sys_tablespaces i 
	WHERE t.table_schema = database() and 
	(i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')) 
	group by t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size
UNION ALL
	SELECT table_name, table_type, unix_timestamp(create_time), table_comment, SUM( data_length + index_length), SUM( data_length + index_length)
	FROM information_schema.tables t
	LEFT OUTER JOIN information_schema.innodb_sys_tablespaces i ON i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')
	WHERE table_schema = database() AND i.name IS NULL
	group by table_name, table_type, unix_timestamp(create_time), table_comment
```

```
169 rows in set (0.74 sec)
```

Next, we can actually replace the `UNION` with a single query:

```sql
	SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, IFNULL(sum(i.file_size), SUM(t.data_length + t.index_length)), IFNULL(sum(i.allocated_size), SUM(t.data_length + t.index_length))
	FROM information_schema.tables t
	LEFT OUTER JOIN information_schema.innodb_sys_tablespaces i ON i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')
	WHERE t.table_schema = database()
	GROUP BY t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size
```

```
169 rows in set (0.37 sec)
```

The next optimization is a bit more exotic. I noticed that individual queries on the `information_schema.tables` and `information_schema.innodb_sys_tablespaces` were fairly quick, but joining them together would cause a quite severe degradation in performance:

```sql
SELECT * FROM information_schema.tables WHERE table_schema = database()
```

```
169 rows in set (0.01 sec)
```

```sql
SELECT * FROM information_schema.innodb_sys_tablespaces WHERE name LIKE CONCAT(database(), '/%');
```

```
Empty set (0.03 sec)
```

So this last optimization step forces early materialization of the data we're interested in:

```sql
SELECT
	t.table_name,
	t.table_type,
	UNIX_TIMESTAMP(t.create_time),
	t.table_comment,
	IFNULL(SUM(i.file_size), SUM(t.data_length + t.index_length)),
	IFNULL(SUM(i.allocated_size), SUM(t.data_length + t.index_length))
FROM information_schema.tables t
LEFT OUTER JOIN (
	SELECT * FROM information_schema.innodb_sys_tablespaces WHERE name LIKE CONCAT(database(), '/%') GROUP BY space
) i ON i.name = CONCAT(t.table_schema, '/', t.table_name) or i.name LIKE CONCAT(t.table_schema, '/', t.table_name, '#p#%')
WHERE t.table_schema = database()
GROUP BY t.table_name
```

```
169 rows in set (0.04 sec)
```

This new query takes only a percent of the time that the old query has taken.

## Checklist
- [ ] Should this PR be backported?
- [ ] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by harshit-gangal** [comment]
Thank you for the contribution. This is nice work.

**Comment by GuptaManan100** [comment]
Looking at each optimization step 1 by 1. 
For the first one, there are multiple articles describing the behaviour of `NOT EXISTS` and `LEFT JOIN ... NULL` being used as anti-joins.  This is one such [article](https://planet.mysql.com/entry/?id=24888) which also suggests that the second alternative will be faster when rows from both sides are non-null, like in this case. To make it even clearer that it is better here are the mysql explain outputs - 
```mysql
mysql [localhost:5731] {msandbox} ((none)) > explain SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, sum(i.file_size), sum(i.allocated_size)
    -> FROM information_schema.tables t, information_schema.innodb_sys_tablespaces i
    -> WHERE t.table_schema = database() and
    -> (i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%'))
    -> group by t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size
    -> UNION ALL
    -> SELECT table_name, table_type, unix_timestamp(create_time), table_comment, SUM( data_length + index_length), SUM( data_length + index_length)
    -> FROM information_schema.tables t
    -> WHERE table_schema = database() AND
    -> NOT EXISTS(SELECT * FROM information_schema.innodb_sys_tablespaces i WHERE i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%'))
    -> group by table_name, table_type, unix_timestamp(create_time), table_comment;
+----+--------------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
| id | select_type        | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                                                                                |
+----+--------------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
|  1 | PRIMARY            | t     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned all databases; Using temporary; Using filesort |
|  1 | PRIMARY            | i     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Using join buffer (Block Nested Loop)                                   |
|  2 | UNION              | t     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned all databases; Using temporary; Using filesort |
|  3 | DEPENDENT SUBQUERY | i     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where                                                                          |
+----+--------------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
4 rows in set, 5 warnings (0.00 sec)
```
and
```mysql
mysql [localhost:5731] {msandbox} ((none)) > explain SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, sum(i.file_size), sum(i.allocated_size)
    -> FROM information_schema.tables t, information_schema.innodb_sys_tablespaces i
    -> WHERE t.table_schema = database() and
    -> (i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%'))
    -> group by t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size
    -> UNION ALL
    -> SELECT table_name, table_type, unix_timestamp(create_time), table_comment, SUM( data_length + index_length), SUM( data_length + index_length)
    -> FROM information_schema.tables t
    -> LEFT OUTER JOIN information_schema.innodb_sys_tablespaces i ON i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')
    -> WHERE table_schema = database() AND i.name IS NULL
    -> group by table_name, table_type, unix_timestamp(create_time), table_comment;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                                                                                |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
|  1 | PRIMARY     | t     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned all databases; Using temporary; Using filesort |
|  1 | PRIMARY     | i     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Using join buffer (Block Nested Loop)                                   |
|  2 | UNION       | t     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned all databases; Using temporary; Using filesort |
|  2 | UNION       | i     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Not exists; Using join buffer (Block Nested Loop)                       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
4 rows in set, 1 warning (0.00 sec)
```

A dependent subquery is being converted to a block nested loop join.

**Comment by GuptaManan100** [comment]
For the second optimization - 
Let's call this predicate `(i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%'))` as `A`;
If a row satisfies predicate A, then `i.name` will be NOT NULL.
The first half of the union is doing a cross join and taking all rows which match the predicate A. It is safe to convert the cross join to left join on predicate A with an additional check that `i.name` is NOT NULL so that no additional rows from the left join qualify

Following this, we now have 2 parts of a UNION, both having a left join on predicate A, the first one selecting rows with `i.name` NOT NULL and the second selecting rows with `i.name` is NULL. 

As long as we don't have rows that qualify from both the sides having the same values for the columns in the group by part, this rewrite should be okay.

**Review by GuptaManan100 — CHANGES_REQUESTED** [review]
The final query does not satisfy FULL GROUP BY mode in MySQL. This is a recommended configuration in Vitess. It fails with the following error - 
```MySql
ERROR 1055 (42000): Expression #2 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'information_schema.innodb_sys_tablespaces.NAME' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by
```
We should make the query such that it obeys full group by. The way to do that is to add the remaining fields to the GROUP BY clauses -
```MySQL
SELECT
	t.table_name,
	t.table_type,
	UNIX_TIMESTAMP(t.create_time),
	t.table_comment,
	IFNULL(SUM(i.file_size), SUM(t.data_length + t.index_length)),
	IFNULL(SUM(i.allocated_size), SUM(t.data_length + t.index_length))
FROM information_schema.tables t
LEFT OUTER JOIN (
	SELECT space, file_size, allocated_size, name FROM information_schema.innodb_sys_tablespaces WHERE name LIKE CONCAT(database(), '/%') GROUP BY space, file_size, allocated_size, name
) i ON i.name = CONCAT(t.table_schema, '/', t.table_name) or i.name LIKE CONCAT(t.table_schema, '/', t.table_name, '#p#%')
WHERE t.table_schema = database()
GROUP BY t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size;
```

**Comment by GuptaManan100** [comment]
After making the query work with FULL GROUP BY mode, I think it will be okay to rewrite it. The plan for the query after step 2 is 
```mysql
mysql [localhost:5731] {msandbox} ((none)) > explain SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, IFNULL(sum(i.file_size), SUM(t.data_length + t.index_length)), IFNULL(sum(i.allocated_size), SUM(t.data_length + t.index_length))
    -> FROM information_schema.tables t
    -> LEFT OUTER JOIN information_schema.innodb_sys_tablespaces i ON i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')
    -> WHERE t.table_schema = database()
    -> GROUP BY t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                                                                                |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
|  1 | SIMPLE      | t     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned all databases; Using temporary; Using filesort |
|  1 | SIMPLE      | i     | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Using join buffer (Block Nested Loop)                                   |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
2 rows in set, 1 warning (0.00 sec)
```
and the plan for the final query will look like - 
```mysql
mysql [localhost:5731] {msandbox} ((none)) > explain SELECT t.table_name, t.table_type, UNIX_TIMESTAMP(t.create_time), t.table_comment, IFNULL(SUM(i.file_size), SUM(t.data_length + t.index_length)), IFNULL(SUM(i.allocated_size), SUM(t.data_length + t.index_length)) FROM information_schema.tables t LEFT OUTER JOIN ( SELECT
space, file_size, allocated_size, name FROM information_schema.innodb_sys_tablespaces WHERE name LIKE CONCAT(database(), '/%') GROUP BY space, file_size, allocate
d_size, name ) i ON i.name = CONCAT(t.table_schema, '/', t.table_name) or i.name LIKE CONCAT(t.table_schema, '/', t.table_name, '#p#%') WHERE t.table_schema = dat
abase() GROUP BY t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, i.file_size;
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
| id | select_type | table                  | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                                                                                |
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
|  1 | PRIMARY     | t                      | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Open_full_table; Scanned all databases; Using temporary; Using filesort |
|  1 | PRIMARY     | <derived2>             | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    2 |   100.00 | Using where; Using join buffer (Block Nested Loop)                                   |
|  2 | DERIVED     | innodb_sys_tablespaces | NULL       | ALL  | NULL          | NULL | NULL    | NULL | NULL |     NULL | Using where; Using temporary; Using filesort                                         |
+----+-------------+------------------------+------------+------+---------------+------+---------+------+------+----------+--------------------------------------------------------------------------------------+
3 rows in set, 1 warning (0.00 sec)
```

**Comment by arthurschreiber** [comment]
@GuptaManan100 the query in this PR was actually fixed to support FULL GROUP BY semantics - I just didn't get around to update the Pull request description as well.

In addition, joining with the subquery does not change the results of the query, but greatly improves performance of it. I can add a comment to the code that explains the reasoning if you feel that this would be useful.

**Review by GuptaManan100 — COMMENTED** [review]
@arthurschreiber Thankyou for the clarification! I think adding a comment would be useful so that we don't rewrite it later. Otherwise everything looks good to me. Just waiting for a second pair of eyes before we merge.

**Review by harshit-gangal — APPROVED** [review]

**Review by systay — APPROVED** [review]
Nice!

**Review by GuptaManan100 — APPROVED** [review]

**Comment by arthurschreiber** [comment]
I extended the comment to mention the reason behind joining a subquery. I also squashed all changes together into a single commit.

**Comment by arthurschreiber** [comment]
Just for reference, loading a database schema that contains 63 `DROP TABLE IF EXISTS` and 63 `CREATE TABLE` statements into a sharded keyspace (2 shards) running on `vtcombo` took 7 minutes 45 seconds.

With this improvement, the same operation takes 50 seconds, while loading the same schema file directly into MySQL takes 16 seconds.

**Review by GuptaManan100 — APPROVED** [review]
This is great! Thankyou @arthurschreiber for your contribution

**Pull request merged**
